libr/bin/demangle.c

```diff
@@ -295,21 +295,87 @@ R_API char *r_bin_demangle_objc(RBinFile *binfile, const char *sym) {
 	return ret;
 }
 
+static bool replace_seq (const char **in, char **out, const char *seq, char value) {
+	size_t len = strlen (seq);
+
+	if (strncmp (*in, seq, len)) {
+		return false;
+	}
+
+	**out = value;
+
+	*in += len;
+	*out += 1;
+
+	return true;
+}
+
+R_API char *r_bin_demangle_rust (RBinFile *binfile, const char *sym, ut64 vaddr) {
+	size_t len;
+	char *str, *out, *in;
+
+	str = r_bin_demangle_cxx (binfile, sym, vaddr);
+	out = in = str;
+	len = strlen (str);
+
+	if (*in == '_') {
+		*in++;
+	}
+
+	while (len-- > 0) {
+		if (!(*in == '$' && (replace_seq (&in, &out, "$SP$", '@')
+				|| replace_seq (&in, &out, "$BP$", '*')
+				|| replace_seq (&in, &out, "$RF$", '&')
+				|| replace_seq (&in, &out, "$LT$", '<')
+				|| replace_seq (&in, &out, "$GT$", '>')
+				|| replace_seq (&in, &out, "$LP$", '(')
+				|| replace_seq (&in, &out, "$RP$", ')')
+				|| replace_seq (&in, &out, "$C$", ',')
+				// maybe a good idea to replace all utf-sequences by regexp \$u[0-9a-f]{2}\$ or so
+				|| replace_seq (&in, &out, "$u20$", ' ')
+				|| replace_seq (&in, &out, "$u22$", '\"')
+				|| replace_seq (&in, &out, "$u27$", '\'')
+				|| replace_seq (&in, &out, "$u2b$", '+')
+				|| replace_seq (&in, &out, "$u3b$", ';')
+				|| replace_seq (&in, &out, "$u5b$", '[')
+				|| replace_seq (&in, &out, "$u5d$", ']')
+				|| replace_seq (&in, &out, "$u7e$", '~')))) {
+			if (*in == '.') {
+				if (in[1] == '.') {
+					in += 2;
+					*out++ = ':';
```

Comment:
(because out can increase faster than in, and the working buffer is just an strdup()